src/vvwe.rs

```diff
@@ -0,0 +1,256 @@
+    pub fn diff_from(&self, other: &Self) -> HashSet<LogTime> {
+        // 1. Find (new) operations that we've seen locally that the remote hasn't
+        let local_ops = (other.max_version.into()..self.max_version.into()).into_iter().filter(|ix : &u64| {
+            !self.exceptions.contains(&(*ix).into())
+        }).map(LogTime::from);
+
+        // 2. Find exceptions that we've seen.
+        let mut local_exceptions  = other.exceptions.difference(&self.exceptions).map(|ix| ix.to_owned());
+
+        local_ops.chain(&mut local_exceptions).collect()
+    }
+}
+
+use crate::vclock::{Actor, Dot};
+
+pub trait CausalOp<A> {
+
+    /// If the result is Some(dot) then this operation cannot occur until the operation that
+    /// occured at dot has.
+    fn happens_after(&self) -> Option<Dot<A>>;
+
+    /// The time that the current operation occured at
+    fn dot(&self) -> Dot<A>;
+}
+
+impl<A: Actor, T: CausalOp<A>> CausalityBarrier<A, T> {
+    pub fn new(site_id: A) -> Self {
+        CausalityBarrier { peers: HashMap::new(), buffer: HashMap::new(), local_id: site_id }
+    }
+
+    pub fn ingest(&mut self, msg: T) -> Option<T> {
+        let v = self.peers.entry(msg.dot().actor).or_insert_with(VectorEntry::new);
+        // Have we already seen this message?
+        if v.is_ready(&msg.dot().counter) {
+            return None
+        }
+
+        v.increment(msg.dot().counter);
+
+        // Ok so it's an exception but maybe we can still integrate it if it's not constrained
+        // by a happens-before relation.
+        // For example: we can always insert into most CRDTs but we can only delete if the
+        // corresponding insert happened before!
+        match msg.happens_after() {
+            // Dang! we have a happens before relation!
+            Some(dot) => {
+                // Let's buffer this operation then.
+                if ! self.saw_site_do(&dot.actor, &dot.counter) {
+                    self.buffer.insert(dot, msg);
+                // and do nothing
+                    None
+                } else {
+                    Some(msg)
+                }
+            }
+            None => {
+                // Ok so we're not causally constrained, but maybe we already saw an associated
+                // causal operation? If so let's just delete the pair
+                match self.buffer.remove(&msg.dot()) {
+                    Some(_) => None,
+                    None => Some(msg),
+                }
+            }
+        }
+    }
+
+    fn saw_site_do(&self, site: &A, t: &LogTime) -> bool {
+        match self.peers.get(site) {
+            Some(ent) => ent.is_ready(t),
+            None => { false }
+        }
+    }
+
+    pub fn expel(&mut self, msg: T) -> T {
+        let v = self.peers.entry(msg.dot().actor).or_insert_with(VectorEntry::new);
+        v.increment(msg.dot().counter);
+        msg
+    }
+
+    pub fn diff_from(&self, other: &HashMap<A, VectorEntry>) -> HashMap<A, HashSet<LogTime>> {
```

Comment:
This is useful to generate "Anti-Entropy" messages which in turn gives you offline support and dynamic join/leave almost for free.

The basic idea is two actors exchange their version vectors, diff them, and gather all the messages they've seen locally that the other doesn't and send them to each other. After this operation both actors will have synchronized their states.

Reply:
very cool